src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -723,27 +722,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             !alloc = bloomFilterAllocForLevel conf ln
             !runPaths = Paths.runPath root (uniqueToRunNumber n)
         traceWith tr $ AtLevel ln $
-          TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergelast
-        let numInputRuns = NumRuns $ V.length rs
-        let numInputEntries = V.foldMap' Run.size rs
+          TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergeLevel
+        -- TODO: creating the merge should happen in MR.new
+        mergeMaybe <- allocateMaybeTemp reg
+          (Merge.new hfs hbio caching alloc mergeLevel resolve runPaths rs)
+          Merge.abort
+        mr <- case mergeMaybe of
+          Nothing -> error "newMerge: merges can not be empty"
+          Just m  -> allocateTemp reg (MR.new mergePolicy rs m) releaseRef
         case confMergeSchedule of
+          Incremental -> pure ()
```

Comment:
Nice. Better factored, less duplication. Make it, and then either do nothing or supply enough credits now to complete in one shot. Nice.